### PR TITLE
add an assert primitive

### PR DESCRIPTION
**Pull request by mattjj**
The assert primitive has an effectful API and so it can't be staged out; it's only a trace-time primitive. It can be discharged to the functional form.

We might want to have separate transforms for discharging errors and for adding error checks. But right now they're just bundled together in the checkify transform.

### CODE REVIEW

File: jax/experimental/checkify.py [hunk 1]
```diff
@@ -212,6 +214,24 @@ def check_errors_traceable(msgs, err, code, *args):
   yield outs
 
 
+## assert primitive
+
+def assert_(pred: Bool, msg: str) -> None:
```
Comment (LenaMartens) [inline jax/experimental/checkify.py:219]: nit: could probably use a nicer public symbol name, but we can think of that later? I guess you're not supposed to overwrite keywords, and we can't provide a `checkify.assert`?

**Review by LenaMartens — APPROVED** [review]
LGTM!

File: jax/experimental/checkify.py (hunk 1 above)
Comment (mattjj) [reply jax/experimental/checkify.py:219]: I think we can't overwrite `assert` actually; it's a syntax error. (I tried, of course!)

You're right that this name is bad. But what's a good one? `jassert`? `assert2`? `asssert`? `jax_assert`? `buttert`? (Brainstormed those with @skye...)

Comment (mattjj) [reply jax/experimental/checkify.py:219]: `allege`?

**Pull request merged**